runtime/tests/integration_test.rs

```diff
@@ -0,0 +1,303 @@
+			assert!(System::events().is_empty());
+			assert_ok!(ParachainStaking::join_candidates(
+				origin_of(AccountId::from(DAVE)),
+				1_000 * GLMR,
+			));
+			assert_eq!(
+				last_event(),
+				Event::parachain_staking(parachain_staking::Event::JoinedCollatorCandidates(
+					AccountId::from(DAVE),
+					1_000 * GLMR,
+					3_100 * GLMR
+				))
+			);
+		});
+}
+
+#[test]
+fn transfer_through_evm_to_stake() {
+	ExtBuilder::default()
+		.with_balances(vec![(AccountId::from(ALICE), 3_000 * GLMR)])
+		.build()
+		.execute_with(|| {
+			// Charlie has no balance => fails to stake
+			assert_noop!(
+				ParachainStaking::join_candidates(
+					origin_of(AccountId::from(CHARLIE)),
+					1_000 * GLMR,
+				),
+				DispatchError::Module {
+					index: 3,
+					error: 3,
+					message: Some("InsufficientBalance")
+				}
+			);
+			// Alice stakes to become a collator candidate
+			assert_ok!(ParachainStaking::join_candidates(
+				origin_of(AccountId::from(ALICE)),
+				1_000 * GLMR,
+			));
+			// Alice transfer from free balance 1000 GLMR to Bob
+			assert_ok!(Balances::transfer(
+				origin_of(AccountId::from(ALICE)),
+				AccountId::from(BOB),
+				2_000 * GLMR,
+			));
+			assert_eq!(Balances::free_balance(AccountId::from(BOB)), 2_000 * GLMR,);
+			use sp_core::U256;
+			let gas_limit = 100000u64;
+			let gas_price: U256 = 1000.into();
+			// Bob transfers 1000 GLMR to Charlie via EVM
+			assert_ok!(Call::EVM(pallet_evm::Call::<Runtime>::call(
+				AccountId::from(BOB),
+				AccountId::from(CHARLIE),
+				Vec::new(),
+				(1_000 * GLMR).into(),
+				gas_limit,
+				gas_price,
+				None
+			))
+			.dispatch(origin_of(AccountId::from(BOB))));
+			assert_eq!(
+				Balances::free_balance(AccountId::from(CHARLIE)),
+				1_000 * GLMR,
+			);
+			// Charlie can stake now
+			assert_ok!(ParachainStaking::join_candidates(
+				origin_of(AccountId::from(CHARLIE)),
+				1_000 * GLMR,
+			),);
+		});
+}
+
+#[test]
+fn reward_block_authors() {
+	ExtBuilder::default()
+		.with_balances(vec![
+			(AccountId::from(ALICE), 2_000 * GLMR),
+			(AccountId::from(BOB), 1_000 * GLMR),
+		])
+		.with_collators(vec![(AccountId::from(ALICE), 1_000 * GLMR)])
+		.with_nominators(vec![(
+			AccountId::from(BOB),
+			AccountId::from(ALICE),
+			500 * GLMR,
+		)])
+		.build()
+		.execute_with(|| {
+			// set parachain inherent data
+			use cumulus_primitives_core::PersistedValidationData;
+			use cumulus_test_relay_sproof_builder::RelayStateSproofBuilder;
+			let (relay_parent_storage_root, relay_chain_state) =
+				RelayStateSproofBuilder::default().into_state_root_and_proof();
+			let vfp = PersistedValidationData {
+				relay_parent_number: 1u32,
+				relay_parent_storage_root,
+				..Default::default()
+			};
+			let parachain_inherent_data = ParachainInherentData {
+				validation_data: vfp,
+				relay_chain_state: relay_chain_state,
+				downward_messages: Default::default(),
+				horizontal_messages: Default::default(),
+			};
+			assert_ok!(Call::ParachainSystem(
```

Comment:
do you mind explaining a little bit more what is happening at this step

Reply:
Definitely, will add comments. We need to set the `relay_chain_block_number` because `author-filter` uses this as a source of randomness to filter valid authors at each block.